src/tensor_parallel/communications.py

```diff
@@ -58,16 +58,45 @@ def __init__(self, world_size: int):
 
 class NCCLAllGather(CollectiveOperation):
     def __init__(self, world_size: int, dim: int):
-        super().__init__(world_size, func=self._nccl_allgather)
+        super().__init__(world_size, func=cross_device_ops.NCCLAllGatherFunction.apply)
+        self.tensor_lengths = [None for _ in range(world_size)]
         self.dim = dim
 
-    def _nccl_allgather(self, *tensors: torch.Tensor):
-        gathered_tensors = cross_device_ops.NCCLAllGatherFunction.apply(*tensors)
-        dim_indices = list(range(1, gathered_tensors[0].ndim))
-        dim_indices.insert(self.dim, 0)
-        concatenated_shape = list(tensors[0].shape)
-        concatenated_shape[self.dim] = -1
-        return tuple(output.permute(dim_indices).reshape(concatenated_shape) for output in gathered_tensors)
+    def __call__(self, x: torch.Tensor, rank: int):
```

Comment:
NCCL's allgather requires tensors to have the same rank; this code circumvents the problem by padding tensors before allgather and un-padding afterwards